enhancements/cloud-integration/openstack/failure-domains.md

```diff
@@ -0,0 +1,720 @@
+* As an administrator, I want to deploy my control plane across multiple
+  independent L2 networks.
+* As an administrator, when I implement failure domains for an OpenStack service
+  that did not previously have them, I want to be able to automatically redeploy
+  my control plane in a running cluster to take advantage of them.
+* As an administrator, when I modify a failure domain within my OpenStack
+  deployment, I want to be able to automatically redeploy my control plane to
+  take advantage of the change.
+* As an administrator who does not use failure domains, I do not want to
+  change any of my existing configuration or tooling due to this change.
+* As a developer of Control Plane MachineSet, I want to know unambiguously what
+  failure domain specification was used for a given Machine so I can ensure the
+  Machine is redeployed if required.
+* As a developer in the Shift on Stack team, I want to ensure that the OpenShift
+  definition of OpenStack failure domains is consistent with the
+  cluster-api-provider-openstack definition of OpenStack failure domains.
+
+### Goals
+
+* Define a unified failure domain specification for use by machines.
+* Integrate this into the installer.
+* Ensure this can be integrated into Control Plane MachineSet.
+
+### Non-Goals
+
+* Integration with Control Plane MachineSet.
+
+## Proposal
+
+In our proposed definition, a failure domain is a bundle of:
+* zero or one Compute zones
+* zero or one Storage zones
+* zero or more ports
+
+We will create a new `OpenStackFailureDomain` type defining a set of Machine
+properties which together ensure the Machine is deployed to a specific,
+consistent failure domain. These properties will be merged into the Machine
+definition by the machine provisioner, which may be either the installer or the
+machine controller.
+
+```go
+type OpenStackFailureDomain struct {
+    // ComputeAvailabilityZone is the name of a valid nova availability zone
+    // +optional
+    ComputeAvailabilityZone string `json:"computeAvailabilityZone,omitempty"`
+
+    // StorageAvailabilityZone is the name of a valid cinder availability zone
+    // +optional
+    StorageAvailabilityZone string `json:"storageAvailabilityZone,omitempty"`
+
+    // Ports defines a set of ports and their attached networks. These will be prepended to any another ports attached to the server.
+    // +optional
+    Ports []PortOpts `json:"ports,omitempty"`
+}
+```
+
+This same struct will be referenced:
+* In the installer's InstallConfig for the initial creation of control plane servers and the creation of relevant manifests.
+* In OpenStackProviderSpec for the creation of a machine using a specific failure domain.
+* In CPMS, for the managed balancing of control plane machines across failure domains.
+
+Leaving any of these fields as their zero value will cause it to be ignored when provisioning the machine. This means a failure domain can define any combination of these fields.
```

Comment:
So if I had two machines, which had providerspecs that look like
```
providerSpec:
  availabilityZone: A
  failureDomain:
    storageAvailabilityZone: A
```
and
```
providerSpec:
  availabilityZone: B
  failureDomain:
    storageAvailabilityZone: A
```
Would you consider these as the same failure domain? If you do, then we cannot use these failure domains within CPMS, since you are varying something outside of your definition of a failure domain.
A failure domain in CPMS must capture all fields that may be different between individual machines in the management of the CPMS.

It strikes me, that if we add this failure domain struct, with the proposed fields, that every other field must be identical outside of the failure domain, so, availability zone and storage availability zone must be the zero value if failure domain is present, right?

In which case, why duplicate those fields into the failure domains struct?